source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,386 @@
+  if (ir_context->get_def_use_mgr()
+          ->GetDef(value_instruction->type_id())
+          ->opcode() != SpvOpTypeInt) {
+    return false;
+  }
+
+  // The size of the integer for value instruction must be equal to 32 bits.
+  auto value_instruction_int_width = ir_context->get_def_use_mgr()
+                                         ->GetDef(value_instruction->type_id())
+                                         ->GetSingleWordInOperand(0);
+
+  if (value_instruction_int_width != 32) {
+    return false;
+  }
+
+  // The first 5 bits of new memory semantics values must be suitable for needed
+  // instruction. The first 5 bits of new value must be larger than the first 5
+  // bits of old. Can't use Sequentially Consistent memory semantic if the
+  // memory model is Vulkan.
+  auto new_memory_sematics_value = value_instruction->GetSingleWordInOperand(0);
+  uint32_t old_memory_sematics_value = 0;
+  if (needed_atomic_instruction->opcode() == SpvOpMemoryBarrier) {
+    // Memory semantics true index for the OpMemoryBarrier equal 1.
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(1))
+            ->GetSingleWordInOperand(0);
+  } else {
+    old_memory_sematics_value =
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                GetMemorySemanticsOperandIndex(
+                    needed_atomic_instruction->opcode(),
+                        message_.memory_semantics_operand_index())))
+            ->GetSingleWordInOperand(0);
+  }
+  auto first_5bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & kMemorySemanticsLowerBitmask);
+  auto first_5bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & kMemorySemanticsLowerBitmask);
+  auto memory_model = static_cast<SpvMemoryModel>(
+      ir_context->module()->GetMemoryModel()->GetSingleWordInOperand(1));
+
+  if (!IsValidConversion(needed_atomic_instruction->opcode(),
+                         first_5bits_old_memory_semantics,
+                         first_5bits_new_memory_semantics, memory_model)) {
+    return false;
+  }
+
+  // The higher bits value of old and new memory semantics id must be equal.
+  auto higher_bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & kMemorySemanticsHigherBitmask);
+  auto higher_bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & kMemorySemanticsHigherBitmask);
+  if (higher_bits_new_memory_semantics != higher_bits_old_memory_semantics) {
+    return false;
+  }
+
+  // Instructions that take two memory semantics and id needed to change are
+  // unequal, the equal id must be stronger than unequal id. Unequal id can't be
+  // released or acquire/release memory semantics.
```

Comment:
I suggest you replace with: Compare and exchange instructions take two memory semantics values: "Equal" and "Unequal". There are extra restrictions in this case.

And then add comments to the statements below, within the `if`, for each check that you are doing.

Reply:
Done.